battle.go

```diff
@@ -143,6 +141,67 @@ func (b *Battle) derefActivePokemon(ap activePokemon) *Pokemon {
 	return (*b.Parties[b.agentParties[ap.AgentIdx]]).Pokemon[ap.PokemonIdx]
 }
 
+// A type that is necessary in order to implement the `Interface` interface, which is used by the sort package to sort.
+type apTurnOrder struct {
+	battle *Battle
+	active []activePokemon
+	turns  map[int]Turn
+	order  []int
+}
+
+func newTurnOrder(battle *Battle, ap []activePokemon, turns map[int]Turn) apTurnOrder {
+	o := []int{}
+	for i := range ap {
+		o = append(o, i)
+	}
+
+	ord := apTurnOrder{
+		battle: battle,
+		order:  o,
+		active: ap,
+		turns:  turns,
+	}
+
+	return ord
+}
+
+func (t *apTurnOrder) GetOrder() []int {
+	return t.order
+}
+
+func (t *apTurnOrder) Len() int {
+	return len(t.order)
+}
+
+func (t *apTurnOrder) Swap(i, j int) {
+	t.order[i], t.order[j] = t.order[j], t.order[i]
+}
+
+// Determine if turn a should happen before turn b.
+func (t *apTurnOrder) Less(a, b int) bool {
+	pA, pB := t.battle.derefActivePokemon(t.active[a]), t.battle.derefActivePokemon(t.active[b])
+	tA, tB := t.turns[a], t.turns[b]
+	if reflect.TypeOf(tA) == reflect.TypeOf(tB) {
+		switch tA.(type) {
+		case FightTurn:
+			// speedy pokemon should go first
+			return pA.Stats[5] > pB.Stats[5]
+		}
+	} else {
+		// make higher priority turns go first
+		return tA.Priority() > tB.Priority()
```

Comment:
This is probably fine for now, but items like the [Quick Claw](https://bulbapedia.bulbagarden.net/wiki/Quick_Claw) will probably cause some additional checks here. My suggestion as mentioned in the meeting is to keep track of some additional flags on Pokemon/in the battle, with quick claw being one of them.

Reply:
Yes, but that check would not go on this line. The priority check here indicates which _type_ of turns should go first. For example, IIRC Item turns should always go before Fight turns.